File: src/main/java/dev/qinx/faker/annotation/datetime/DateTime.java
package dev.qinx.faker.annotation.datetime;

import dev.qinx.faker.provider.Provider;
import dev.qinx.faker.provider.datetime.LocalDateTimeProvider;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * Generate a LocalDateTime object with a given format for the annotated field. This annotation could also be used
 * on a String field.
 *
 * <p>The default format is the ISO-8601 extended date-time format, like "2011-12-03T10:15:30"</p>
 */
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.FIELD})
public @interface DateTime {
    String format() default "";

    String seed() default "";

    Class<? extends Provider> provider() default LocalDateTimeProvider.class;
}

File: src/main/scala/dev/qinx/faker/provider/datetime/Datetime.scala
package dev.qinx.faker.provider.datetime

import java.time.{LocalDate, LocalDateTime, LocalTime}

import dev.qinx.faker.Faker

class Datetime(val seed: Option[Long] = None) {

  def localDate(): LocalDate = {
    val providerID = s"LocalDateProvider${seed.getOrElse("")}"
    Faker.provide[LocalDate](providerID) {
      id => new LocalDateProvider().setSeed(seed).setProviderID(id)
    }
  }

  def localDateTime(): LocalDateTime = {
    val providerID = s"LocalDateTimeProvider${seed.getOrElse("")}"
    Faker.provide[LocalDateTime](providerID) {
      id => new LocalDateTimeProvider().setSeed(seed).setProviderID(id)
    }
  }

  def localTime(): LocalTime = {
    val providerID = s"LocalTimeProvider${seed.getOrElse("")}"
    Faker.provide[LocalTime](providerID) {
      id => new LocalTimeProvider().setSeed(seed).setProviderID(id)
    }
  }

}
